File: src/argentum/java/org/taumc/celeritas/impl/render/terrain/fog/KaliaFogService.java
package org.taumc.celeritas.impl.render.terrain.fog;

import org.embeddedt.embeddium.impl.render.chunk.fog.FogService;
import org.embeddedt.embeddium.impl.render.chunk.shader.ChunkFogMode;
import re.lilith.kalia.gl.ShaderUniforms;

public class KaliaFogService implements FogService {
    @Override
    public float getFogEnd() {
        return ShaderUniforms.INSTANCE.fogEnd();
    }

    @Override
    public float getFogStart() {
        return ShaderUniforms.INSTANCE.fogStart();
    }

    @Override
    public float getFogDensity() {
        return ShaderUniforms.INSTANCE.fogDensity();
    }

    @Override
    public int getFogShapeIndex() {
        return 0;
    }

    @Override
    public float getFogCutoff() {
        return getFogEnd();
    }

    @Override
    public float[] getFogColor() {
        return new float[]{ShaderUniforms.INSTANCE.fogRed(), ShaderUniforms.INSTANCE.fogGreen(), ShaderUniforms.INSTANCE.fogBlue(), 1.0F};
    }

    @Override
    public ChunkFogMode getFogMode() {
        if (!ShaderUniforms.INSTANCE.isFogEnabled()) {
            return ChunkFogMode.NONE;
        }
        return ChunkFogMode.fromGLMode(ShaderUniforms.INSTANCE.fogMode().getGlMode());
    }
}

File: src/main/kotlin/re/lilith/kalia/shader/fog/KaliaFogService.kt
package re.lilith.kalia.shader.fog

import org.embeddedt.embeddium.impl.render.chunk.fog.FogService
import org.embeddedt.embeddium.impl.render.chunk.shader.ChunkFogMode
import re.lilith.kalia.gl.ShaderUniforms.fogBlue
import re.lilith.kalia.gl.ShaderUniforms.fogDensity
import re.lilith.kalia.gl.ShaderUniforms.fogEnd
import re.lilith.kalia.gl.ShaderUniforms.fogGreen
import re.lilith.kalia.gl.ShaderUniforms.fogMode
import re.lilith.kalia.gl.ShaderUniforms.fogRed
import re.lilith.kalia.gl.ShaderUniforms.fogStart
import re.lilith.kalia.gl.ShaderUniforms.isFogEnabled

class KaliaFogService : FogService {
    override fun getFogEnd(): Float {
        return fogEnd()
    }

    override fun getFogStart(): Float {
        return fogStart()
    }

    override fun getFogDensity(): Float {
        return fogDensity()
    }

    override fun getFogShapeIndex(): Int {
        return 0
    }

    override fun getFogCutoff(): Float {
        return getFogEnd()
    }

    override fun getFogColor(): FloatArray? {
        return floatArrayOf(fogRed(), fogGreen(), fogBlue(), 1.0f)
    }

    override fun getFogMode(): ChunkFogMode? {
        if (!isFogEnabled()) {
            return ChunkFogMode.NONE
        }
        return ChunkFogMode.fromGLMode(fogMode().glMode)
    }
}
